crypto/context.c

```diff
@@ -299,127 +514,114 @@ int ossl_lib_ctx_is_global_default(OSSL_LIB_CTX *ctx)
+    case OSSL_LIB_CTX_DRBG_NONCE_INDEX:
+        return ctx->drbg_nonce;
+#ifndef FIPS_MODULE
+    case OSSL_LIB_CTX_PROVIDER_CONF_INDEX:
+        return ctx->provider_conf;
+    case OSSL_LIB_CTX_BIO_CORE_INDEX:
+        return ctx->bio_core;
+    case OSSL_LIB_CTX_CHILD_PROVIDER_INDEX:
+        return ctx->child_provider;
+    case OSSL_LIB_CTX_DECODER_STORE_INDEX:
+        return ctx->decoder_store;
+    case OSSL_LIB_CTX_ENCODER_STORE_INDEX:
+        return ctx->encoder_store;
+    case OSSL_LIB_CTX_STORE_LOADER_STORE_INDEX:
+        return ctx->store_loader_store;
+    case OSSL_LIB_CTX_SELF_TEST_CB_INDEX:
+        return ctx->self_test_cb;
+#endif
 
-    if (dynidx != -1) {
-        if (!CRYPTO_THREAD_read_lock(ctx->index_locks[index]))
-            return NULL;
-        if (!CRYPTO_THREAD_read_lock(ctx->lock)) {
-            CRYPTO_THREAD_unlock(ctx->index_locks[index]);
+    case OSSL_LIB_CTX_RAND_CRNGT_INDEX: {
+
+        /*
+         * rand_crngt must be lazily initialized because it calls into
+         * libctx, so must not be called from context_init, else a deadlock
+         * will occur.
+         *
+         * We use a separate lock because code called by the instantiation
+         * of rand_crngt is liable to try and take the libctx lock.
+         */
+        if (CRYPTO_THREAD_read_lock(ctx->rand_crngt_lock) != 1)
             return NULL;
+
+        if (ctx->rand_crngt == NULL) {
+            CRYPTO_THREAD_unlock(ctx->rand_crngt_lock);
+
+            if (CRYPTO_THREAD_write_lock(ctx->rand_crngt_lock) != 1)
+                return NULL;
+
+            if (ctx->rand_crngt == NULL)
+                ctx->rand_crngt = ossl_rand_crng_ctx_new(ctx);
         }
-        data = CRYPTO_get_ex_data(&ctx->data, dynidx);
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return data;
-    }
 
-    if (!CRYPTO_THREAD_write_lock(ctx->index_locks[index]))
-        return NULL;
-    if (!CRYPTO_THREAD_write_lock(ctx->lock)) {
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return NULL;
-    }
+        p = ctx->rand_crngt;
 
-    dynidx = ctx->dyn_indexes[index];
-    if (dynidx != -1) {
-        data = CRYPTO_get_ex_data(&ctx->data, dynidx);
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return data;
-    }
+        CRYPTO_THREAD_unlock(ctx->rand_crngt_lock);
 
-    if (!ossl_lib_ctx_init_index(ctx, index, meth)) {
-        CRYPTO_THREAD_unlock(ctx->lock);
-        CRYPTO_THREAD_unlock(ctx->index_locks[index]);
-        return NULL;
+        return p;
     }
 
-    CRYPTO_THREAD_unlock(ctx->lock);
+#ifdef FIPS_MODULE
+    case OSSL_LIB_CTX_THREAD_EVENT_HANDLER_INDEX:
+        return ctx->thread_event_handler;
+#endif
+
+    case OSSL_LIB_CTX_FIPS_PROV_INDEX: {
+        /*
+         * fips is a separate module which may or may not be loaded,
+         * so we have to do this lazily.
+         */
```

Comment:
I'm not sure why it isn't sufficient to do this eagerly but wrapped in "ifdef FIPS_MODULE".

Reply:
As I recall, trying to initialize the FIPS module eagerly resulted in a rather obtuse issue caused by FIPS trying to call into libctx methods before libctx_init had been completed.